docs/design/DAM-on-type-and-RUC-interaction.md

```diff
@@ -0,0 +1,151 @@
+{
+    [RUC]
+    public void MethodWithRUC() {}
+
+    public void MethodWithDAM([DAM(Fields)] Type type) {}
+}
+```
+
+Following the existing rules, accessing `MethodWithRUC` or `MethodWithDAM` via reflection will generate a warning, regardless of the RUC on type.
+Since DAM on type is meant to behave as if accessing the members via reflection, those warnings should not be affected by the RUC on type either.
+
+## RUC on method and interaction with DAM on type
+
+```csharp
+[DAM(All)]
+class AnnotatedBase {}
+
+class Derived : AnnotatedBase
+{
+    [RUC]
+    public void MethodWithRUC() {}
+
+    public void MethodWithDAM([DAM(Fields)] Type type) {}
+
+    [RUC]
+    public void MethodWithRUCAndDAM([DAM(Fields)] Type type) {}
+}
+```
+
+Following the existing rules, accessing `MethodWithRUC` or `MethodWithDAM` via reflection will generate a warning. The same rules also imply that accessing `MethodWithRUCAndDAM` via reflection will also generate a warning (actually two of them).
+RUC on method does not affect usage of the method (other than it should generate a warning on the usage).
+
+## RUC or DAM on nested type
+
+```csharp
+[DAM(All)]
+class AnnotatedBase {}
+
+class Derived : AnnotatedBase
+{
+    [RUC]
+    class NestedWithRUC {}
+
+    [DAM]
+    class NestedWithDAM {}
+}
+```
+
+This is probably an existing hole. Accessing a RUC or DAM type itself via reflection is not problematic on its own (since both annotations should behave as applying to members only). But it can create a hole in the analysis because of annotation propagation through `GetNestedType`. For example (given the above code):
+
+```csharp
+void TestMethod(Derived instance)
+{
+    Type typeOfDerived = instance.GetType(); // The type value is effectively annotated with All
+    Type nestedWithRUC = typeOfDerived.GetNestedType("NestedWithRUC"); // The resulting type value is effectively annotated with All
+    Activator.CreateInstance(nestedWithRUC); // No warning - nestedWithRUC has All annotation on it
+}
+```
+
+The above MUST warn, since it's calling a `.ctor` on a RUC annotated type.
+
+DAM on nested type doesn't seem to have a problem mainly because for marking purposes it's not needed (the DAM on the outer type if it applies to the nested type will mark the entire nested type, so DAM on it can't mark more).
+Also accessing DAM annotated types themselves is not a problem, and accessing their members is drive by the DAM annotation.
+This needs verification!!! (It is tricky)
+
+## Conclusion
+
+Personally I'm not a big fan of the fact that RUC on type doesn't suppress the new warnings on the type's members - as a normal user I would expect that to happen. That said, doing so would introduce an "Exception" to the existing rules, which is not ideal either. If we warn more (RUC on type doesn't affect the new warnings), it sort of creates noise, but it's not exactly wrong. And we can change our mind later on (by not warning anymore some day in the future).
```

Comment:
What sold me on the proposed behavior is:
```C#
[DAM(All)]
class Annotated
{
    [RUC]
    public void MethodWithRUC() {}
}
```

If RUC was to suppress all warnings about a method, it should also suppress the `IL2112` in this case, but that's nonsense, as the RUC would effectively suppress "itself".

I think this is basically the problem of marking because of two unrelated locations. The warning is basically saying "Something somewhere may use this method through reflection - and that's a problem". I think that the current experience is good - the thing to blame is the combination of the type having annotation and the method being on that type. We will point out both (the method and the type) in the warning message - and putting the warning onto the method provides much better source-code location (and makes it much more precise to suppress).

Reply:
Yup, the argument about more precise suppressions is the most convincing argument to me. I was just pointing out how the member warning about something else accessing it could be confusing since it's usually the other way around.